metacat-metadata-mysql/src/main/java/com/netflix/metacat/metadata/mysql/MySqlParentChildRelMetaDataService.java

```diff
@@ -0,0 +1,229 @@
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.sql.PreparedStatement;
+import java.util.Set;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.stream.Collectors;
+
+/**
+ * Parent Child Relationship Metadata Service.
+ * This stores the parent child relationship of two entities as first class citizen in Metacat.
+ */
+@Slf4j
+@SuppressFBWarnings
+@Transactional("metadataTxManager")
+public class MySqlParentChildRelMetaDataService implements ParentChildRelMetadataService {
+    static final String SQL_CREATE_PARENT_CHILD_RELATIONS =
+        "INSERT INTO parent_child_relation (parent, parent_uuid, child, child_uuid, relation_type) "
+            + "VALUES (?, ?, ?, ?, ?)";
+
+    static final String SQL_DELETE_PARENT_CHILD_RELATIONS =
+        "DELETE FROM parent_child_relation "
+            + "WHERE parent = ? AND parent_uuid = ? AND child = ? AND child_uuid = ? AND relation_type = ?";
+
+    static final String SQL_RENAME_PARENT_ENTITY = "UPDATE parent_child_relation "
+        + "SET parent = ? WHERE parent = ?";
+    static final String SQL_RENAME_CHILD_ENTITY = "UPDATE parent_child_relation "
+        + "SET child = ? WHERE child = ?";
+
+    static final String SQL_DROP_CHILD = "DELETE FROM parent_child_relation "
+        + "WHERE child = ? ";
+    static final String SQL_DROP_PARENT = "DELETE FROM parent_child_relation "
+        + "WHERE parent = ? ";
+
+    static final String SQL_GET_PARENTS = "SELECT parent, parent_uuid, relation_type "
+        + "FROM parent_child_relation WHERE child = ?";
+
+    static final String SQL_GET_CHILDREN = "SELECT child, child_uuid, relation_type "
+        + "FROM parent_child_relation WHERE parent = ?";
+
+    private final JdbcTemplate jdbcTemplate;
+    private final ConverterUtil converterUtil;
+
+    /**
+     * Constructor.
+     *
+     * @param jdbcTemplate jdbc template
+     * @param converterUtil converterUtil
+     */
+    @Autowired
+    public MySqlParentChildRelMetaDataService(final JdbcTemplate jdbcTemplate, final ConverterUtil converterUtil) {
+        this.jdbcTemplate = jdbcTemplate;
+        this.converterUtil = converterUtil;
+    }
+
+    @Override
+    public void createParentChildRelation(final QualifiedName parentName,
+                                          final String parentUUID,
+                                          final QualifiedName childName,
+                                          final String childUUID,
+                                          final String type) {
+        // Validation to prevent having a child have two parents
+        final Set<ParentInfo> childParents = getParents(childName);
```

Comment:
we can move this enforcement to the db with:
```
CREATE TABLE `parent_child_relation` (
    `parent` varchar(255) NOT NULL,
    `parent_uuid` varchar(255) NOT NULL,
    `child` varchar(255) NOT NULL,
    `child_uuid` varchar(255) NOT NULL,
    `relation_type` varchar(255) NOT NULL,
    PRIMARY KEY (`parent`, `child`, `parent_uuid`, `child_uuid`, `relation_type`),
    UNIQUE KEY `unique_child` (`child`, `child_uuid`),
    INDEX `idx_child` (`child`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1;
```

but this would break our new rename logic where we have that temporary stage with both old and new relationship. we need to handle that either way in this method, so perhaps we dont enforce at the db and we pass down a `isRename` arg to this method

Reply:
I agree but then the error message we populate might not be clear because it might break the primary key unique or the unique key and if we define them in db level, it's hard to tell which key was violated.